From 16ffddf0a8c2c8198fd6183337511e0428ff2051 Mon Sep 17 00:00:00 2001
From: Eric Lundby <Eric.Lundby@gmail.com>
Date: Thu, 22 Jan 2026 11:13:42 -0700
Subject: [PATCH] Create abseil systemlib patch for TensorFlow

Creates abseil_systemlib.patch that gets applied to the TensorFlow source
during Bazel build, wiring the XLA absl targets to conda's system abseil.
Regenerated against TensorFlow 2.21.0 (the previous 2.20-era content no
longer applied).
---
 third_party/tensorflow/BUILD                  |   3 +
 third_party/tensorflow/abseil_systemlib.patch | 421 ++++++++++++++++++++++++++
 2 files changed, 424 insertions(+)

diff --git a/third_party/tensorflow/BUILD b/third_party/tensorflow/BUILD
index 2b0003e..5c63f95 100644
--- a/third_party/tensorflow/BUILD
+++ b/third_party/tensorflow/BUILD
@@ -1 +1,4 @@
 # Needed for Bazel to treat this directory as a package
+
+# Export patch files for use by http_archive rules
+exports_files(["abseil_systemlib.patch"])
diff --git a/third_party/tensorflow/abseil_systemlib.patch b/third_party/tensorflow/abseil_systemlib.patch
new file mode 100644
index 0000000..887866c
--- /dev/null
+++ b/third_party/tensorflow/abseil_systemlib.patch
@@ -0,0 +1,421 @@
+diff --git a/third_party/xla/third_party/absl/system.absl.base.BUILD b/third_party/xla/third_party/absl/system.absl.base.BUILD
+index d6bf8748..f88bef77 100644
+--- a/third_party/xla/third_party/absl/system.absl.base.BUILD
++++ b/third_party/xla/third_party/absl/system.absl.base.BUILD
+@@ -13,6 +13,7 @@ package(default_visibility = ["//visibility:public"])
+     "errno_saver",
+     "fast_type_id",
+     "pretty_function",
++    "no_destructor",
+ ]]
+ 
+ cc_library(
+@@ -64,6 +65,23 @@ cc_library(
+     ],
+ )
+ 
++cc_library(
++    name = "nullability",
++    defines = [
++        "absl_nonnull=",
++        "absl_nullable=", 
++        "absl_nullability_unknown=",
++    ],
++)
++
++cc_library(
++    name = "prefetch",
++    deps = [
++        ":config",
++        ":core_headers",
++    ],
++)
++
+ cc_library(
+     name = "throw_delegate",
+     linkopts = ["-labsl_throw_delegate"],
+@@ -105,3 +123,12 @@ cc_library(
+         "//absl:__subpackages__",
+     ],
+ )
++
++cc_library(
++    name = "tracing_internal",
++    linkopts = ["-labsl_base", "-labsl_tracing_internal"],
++    deps = [
++        ":config",
++        ":core_headers",
++    ],
++)
+diff --git a/third_party/xla/third_party/absl/system.absl.container.BUILD b/third_party/xla/third_party/absl/system.absl.container.BUILD
+index 95c16260..ba5962be 100644
+--- a/third_party/xla/third_party/absl/system.absl.container.BUILD
++++ b/third_party/xla/third_party/absl/system.absl.container.BUILD
+@@ -215,3 +215,11 @@ cc_library(
+         "//absl/utility",
+     ],
+ )
++
++cc_library(
++    name = "hash_container_defaults",
++    deps = [
++        ":hash_function_defaults",
++        "//absl/base:config",
++    ],
++)
+diff --git a/third_party/xla/third_party/absl/system.absl.debugging.BUILD b/third_party/xla/third_party/absl/system.absl.debugging.BUILD
+index 931ffdc9..3fc97a45 100644
+--- a/third_party/xla/third_party/absl/system.absl.debugging.BUILD
++++ b/third_party/xla/third_party/absl/system.absl.debugging.BUILD
+@@ -4,7 +4,7 @@ package(default_visibility = ["//visibility:public"])
+ 
+ cc_library(
+     name = "stacktrace",
+-    linkopts = ["-labsl_stacktrace"],
++    linkopts = ["-labsl_stacktrace", "-labsl_examine_stack"],
+     deps = [
+         ":debugging_internal",
+     ],
+@@ -41,7 +41,7 @@ cc_library(
+ 
+ cc_library(
+     name = "debugging_internal",
+-    linkopts = ["-labsl_debugging_internal"],
++    linkopts = ["-labsl_stacktrace"],
+     deps = [
+         "//absl/base:dynamic_annotations",
+         "//absl/base:errno_saver",
+@@ -51,7 +51,7 @@ cc_library(
+ 
+ cc_library(
+     name = "demangle_internal",
+-    linkopts = ["-labsl_demangle_internal"],
++    linkopts = ["-labsl_symbolize"],
+     deps = [
+         "//absl/base",
+     ],
+diff --git a/third_party/xla/third_party/absl/system.absl.flags.BUILD b/third_party/xla/third_party/absl/system.absl.flags.BUILD
+index aff653c7..2378fe2b 100644
+--- a/third_party/xla/third_party/absl/system.absl.flags.BUILD
++++ b/third_party/xla/third_party/absl/system.absl.flags.BUILD
+@@ -5,9 +5,7 @@ package(default_visibility = ["//visibility:public"])
+ cc_library(
+     name = "program_name",
+     linkopts = ["-labsl_flags_program_name"],
+-    visibility = [
+-        "//absl/flags:__pkg__",
+-    ],
++    visibility = ["//visibility:public"],
+     deps = [
+         "//absl/strings",
+         "//absl/synchronization",
+@@ -36,7 +34,7 @@ cc_library(
+ 
+ cc_library(
+     name = "commandlineflag_internal",
+-    linkopts = ["-labsl_flags_commandlineflag_internal"],
++    linkopts = ["-labsl_flags_commandlineflag", "-labsl_flags_commandlineflag_internal"],
+ )
+ 
+ cc_library(
+@@ -52,9 +50,7 @@ cc_library(
+ cc_library(
+     name = "private_handle_accessor",
+     linkopts = ["-labsl_flags_private_handle_accessor"],
+-    visibility = [
+-        "//absl/flags:__pkg__",
+-    ],
++    visibility = ["//visibility:public"],
+     deps = [
+         ":commandlineflag",
+         ":commandlineflag_internal",
+@@ -79,7 +75,7 @@ cc_library(
+ cc_library(
+     name = "flag_internal",
+     linkopts = ["-labsl_flags_internal"],
+-    visibility = ["//absl/base:__subpackages__"],
++    visibility = ["//visibility:public"],
+     deps = [
+         ":commandlineflag",
+         ":commandlineflag_internal",
+@@ -97,11 +93,24 @@ cc_library(
+ 
+ cc_library(
+     name = "flag",
+-    linkopts = ["-labsl_flags"],
++    linkopts = [
++        "-labsl_flags_internal",
++        "-labsl_flags_commandlineflag",
++        "-labsl_flags_config",
++        "-labsl_flags_marshalling",
++        "-labsl_flags_reflection",
++        "-labsl_flags_private_handle_accessor",
++        "-labsl_flags_program_name",
++    ],
+     deps = [
+         ":config",
+         ":flag_internal",
+         ":reflection",
++        ":commandlineflag",
++        ":commandlineflag_internal",
++        ":marshalling",
++        ":private_handle_accessor",
++        ":program_name",
+         "//absl/base",
+         "//absl/strings",
+     ],
+@@ -109,10 +118,8 @@ cc_library(
+ 
+ cc_library(
+     name = "usage_internal",
+-    linkopts = ["-labsl_flags_usage_internal"],
+-    visibility = [
+-        "//absl/flags:__pkg__",
+-    ],
++    linkopts = ["-labsl_flags_usage"],
++    visibility = ["//visibility:public"],
+     deps = [
+         ":commandlineflag",
+         ":config",
+diff --git a/third_party/xla/third_party/absl/system.absl.functional.BUILD b/third_party/xla/third_party/absl/system.absl.functional.BUILD
+index 9439bd0b..a12a4466 100644
+--- a/third_party/xla/third_party/absl/system.absl.functional.BUILD
++++ b/third_party/xla/third_party/absl/system.absl.functional.BUILD
+@@ -13,3 +13,8 @@ cc_library(
+ cc_library(
+     name = "function_ref",
+ )
++
++cc_library(
++    name = "overload",
++    linkopts = ["-labsl_base"],
++)
+\ No newline at end of file
+diff --git a/third_party/xla/third_party/absl/system.absl.hash.BUILD b/third_party/xla/third_party/absl/system.absl.hash.BUILD
+index 3367340c..316693e1 100644
+--- a/third_party/xla/third_party/absl/system.absl.hash.BUILD
++++ b/third_party/xla/third_party/absl/system.absl.hash.BUILD
+@@ -28,7 +28,7 @@ cc_library(
+ 
+ cc_library(
+     name = "low_level_hash",
+-    linkopts = ["-labsl_low_level_hash"],
++    linkopts = ["-labsl_hash"],
+     visibility = ["//visibility:private"],
+     deps = [
+         "//absl/base:endian",
+diff --git a/third_party/xla/third_party/absl/system.absl.random.BUILD b/third_party/xla/third_party/absl/system.absl.random.BUILD
+index ac17ce63..f3a4bcfc 100644
+--- a/third_party/xla/third_party/absl/system.absl.random.BUILD
++++ b/third_party/xla/third_party/absl/system.absl.random.BUILD
+@@ -29,14 +29,13 @@ cc_library(
+ cc_library(
+     name = "seed_sequences",
+     linkopts = [
+-        "-labsl_random_internal_platform",
+-        "-labsl_random_internal_pool_urbg",
++        "-labsl_random_seed_sequences",
++        "-labsl_random_internal_entropy_pool",
+         "-labsl_random_internal_randen",
+         "-labsl_random_internal_randen_hwaes",
+         "-labsl_random_internal_randen_hwaes_impl",
+         "-labsl_random_internal_randen_slow",
+-        "-labsl_random_internal_seed_material",
+-        "-labsl_random_seed_sequences",
++        "-labsl_random_internal_platform",
+         "-pthread",
+     ],
+     deps = [
+diff --git a/third_party/xla/third_party/absl/system.absl.status.BUILD b/third_party/xla/third_party/absl/system.absl.status.BUILD
+index e50e9790..3eeb0d8d 100644
+--- a/third_party/xla/third_party/absl/system.absl.status.BUILD
++++ b/third_party/xla/third_party/absl/system.absl.status.BUILD
+@@ -4,7 +4,14 @@ package(default_visibility = ["//visibility:public"])
+ 
+ cc_library(
+     name = "status",
+-    linkopts = ["-labsl_status"],
++    linkopts = [
++        "-labsl_status",
++        "-labsl_cord",
++        "-labsl_cord_internal",
++        "-labsl_cordz_functions",
++        "-labsl_cordz_info",
++        "-labsl_cordz_handle",
++    ],
+     deps = [
+         "//absl/base:atomic_hook",
+         "//absl/base:raw_logging_internal",
+diff --git a/third_party/xla/third_party/absl/system.absl.strings.BUILD b/third_party/xla/third_party/absl/system.absl.strings.BUILD
+index fa9a7a84..f7fa5737 100644
+--- a/third_party/xla/third_party/absl/system.absl.strings.BUILD
++++ b/third_party/xla/third_party/absl/system.absl.strings.BUILD
+@@ -17,16 +17,27 @@ cc_library(
+ 
+ cc_library(
+     name = "internal",
+-    linkopts = ["-labsl_strings_internal"],
++    linkopts = ["-labsl_strings", "-labsl_strings_internal"],
+     deps = [
+         "//absl/base:endian",
+         "//absl/base:raw_logging_internal",
+     ],
+ )
+ 
++cc_library(
++    name = "string_view",
++    deps = [
++        "//absl/base",
++        "//absl/base:config",
++        "//absl/base:core_headers",
++        "//absl/base:nullability",
++        "//absl/base:throw_delegate",
++    ],
++)
++
+ cc_library(
+     name = "cord",
+-    linkopts = ["-labsl_cord"],
++    linkopts = ["-labsl_cord", "-labsl_cordz_functions", "-labsl_cordz_info", "-labsl_cordz_handle", "-labsl_cord_internal"],
+     deps = [
+         ":str_format",
+         "//absl/container:compressed_tuple",
+diff --git a/third_party/xla/third_party/absl/system.absl.synchronization.BUILD b/third_party/xla/third_party/absl/system.absl.synchronization.BUILD
+index c0fa37aa..875b5bca 100644
+--- a/third_party/xla/third_party/absl/system.absl.synchronization.BUILD
++++ b/third_party/xla/third_party/absl/system.absl.synchronization.BUILD
+@@ -5,7 +5,7 @@ package(default_visibility = ["//visibility:public"])
+ # Internal data structure for efficiently detecting mutex dependency cycles
+ cc_library(
+     name = "graphcycles_internal",
+-    linkopts = ["-labsl_graphcycles_internal"],
++    linkopts = ["-labsl_synchronization"],
+     visibility = [
+         "//absl:__subpackages__",
+     ],
+@@ -16,6 +16,20 @@ cc_library(
+     ],
+ )
+ 
++cc_library(
++    name = "kernel_timeout_internal",
++    linkopts = ["-labsl_kernel_timeout_internal"],
++    visibility = [
++    ],
++    deps = [
++        "//absl/base",
++        "//absl/base:config",
++        "//absl/base:core_headers",
++        "//absl/base:raw_logging_internal",
++        "//absl/time",
++    ],
++)
++
+ cc_library(
+     name = "synchronization",
+     linkopts = [
+@@ -24,6 +38,7 @@ cc_library(
+     ],
+     deps = [
+         ":graphcycles_internal",
++        ":kernel_timeout_internal",
+         "//absl/base",
+         "//absl/base:atomic_hook",
+         "//absl/base:dynamic_annotations",
+diff --git a/third_party/xla/third_party/absl/system.absl.types.BUILD b/third_party/xla/third_party/absl/system.absl.types.BUILD
+index db94fc99..ebfea47b 100644
+--- a/third_party/xla/third_party/absl/system.absl.types.BUILD
++++ b/third_party/xla/third_party/absl/system.absl.types.BUILD
+@@ -11,7 +11,9 @@ cc_library(
+ 
+ cc_library(
+     name = "bad_any_cast",
+-    linkopts = ["-labsl_bad_any_cast_impl"],
++    deps = [
++        "@com_google_absl//absl/base:base",
++    ],
+ )
+ 
+ cc_library(
+@@ -30,16 +32,16 @@ cc_library(
+ 
+ cc_library(
+     name = "bad_optional_access",
+-    linkopts = ["-labsl_bad_optional_access"],
+     deps = [
++        "@com_google_absl//absl/base:base",
+         "//absl/base:raw_logging_internal",
+     ],
+ )
+ 
+ cc_library(
+     name = "bad_variant_access",
+-    linkopts = ["-labsl_bad_variant_access"],
+     deps = [
++        "@com_google_absl//absl/base:base",
+         "//absl/base:raw_logging_internal",
+     ],
+ )
+diff --git a/third_party/xla/third_party/absl/system.absl.utility.BUILD b/third_party/xla/third_party/absl/system.absl.utility.BUILD
+index e15049e2..e9c25fe3 100644
+--- a/third_party/xla/third_party/absl/system.absl.utility.BUILD
++++ b/third_party/xla/third_party/absl/system.absl.utility.BUILD
+@@ -1,6 +1,11 @@
+ load("@rules_cc//cc:defs.bzl", "cc_library")
+ 
++package(default_visibility = ["//visibility:public"])
++
+ cc_library(
+     name = "utility",
+-    visibility = ["//visibility:public"],
+ )
++
++cc_library(
++    name = "if_constexpr",
++)
+\ No newline at end of file
+diff --git a/third_party/xla/third_party/absl/workspace.bzl b/third_party/xla/third_party/absl/workspace.bzl
+index 46c1573b..6a6add09 100644
+--- a/third_party/xla/third_party/absl/workspace.bzl
++++ b/third_party/xla/third_party/absl/workspace.bzl
+@@ -11,9 +11,40 @@ def repo():
+     ABSL_SHA256 = "d1abe9da2003e6cbbd7619b0ced3e52047422f4f4ac6c66a9bef5d2e99fea837"
+     # LINT.ThenChange(//tensorflow/lite/tools/cmake/modules/abseil-cpp.cmake)
+ 
++    SYS_DIRS = [
++        "algorithm",
++        "base",
++        "crc",
++        "log",
++        "cleanup",
++        "container",
++        "debugging",
++        "flags",
++        "functional",
++        "hash",
++        "memory",
++        "meta",
++        "numeric",
++        "random",
++        "status",
++        "strings",
++        "synchronization",
++        "time",
++        "types",
++        "utility",
++    ]
++
++    SYS_LINKS = {
++        "//third_party/absl:system.absl.{name}.BUILD".format(name = n): "absl/{name}/BUILD.bazel".format(name = n)
++        for n in SYS_DIRS
++    }
++
+     tf_http_archive(
+         name = "com_google_absl",
+         sha256 = ABSL_SHA256,
++        build_file = "//third_party/absl:com_google_absl.BUILD",
++        system_build_file = "//third_party/absl:system.BUILD",
++        system_link_files = SYS_LINKS,
+         strip_prefix = "abseil-cpp-{commit}".format(commit = ABSL_COMMIT),
+         urls = tf_mirror_urls("https://github.com/abseil/abseil-cpp/archive/{commit}.tar.gz".format(commit = ABSL_COMMIT)),
+         patch_file = [
